core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```diff
@@ -96,6 +114,7 @@ private[spark] object JsonProtocol {
         executorMetricsUpdateToJson(metricsUpdate)
       case blockUpdated: SparkListenerBlockUpdated =>
         throw new MatchError(blockUpdated)  // TODO(ekl) implement this
+      case _ => parse(write(event))
```

Comment:
I tried using the `JsonTypeInfo` annotation and it worked. The difference is that it need write the full class name now. The event will look like 
```
"Event":"org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionStart"
```
When deserializing the data, we use reflection to find the class so that the class list is not needed.